### PR TITLE
add track by statement to all ng-repeat attributes

### PR DESCRIPTION
**Pull request by Andrux51**
Going off of information seen in [PR#3993](https://github.com/HabitRPG/habitrpg/pull/3993), this adds a track by statement to every instance of the ngRepeat directive. Hoping this makes noticeable impact on Issue #3486 - screens with large numbers of chat messages are slowing to extreme levels, notably the chat box itself is working slowly.

I did some cursory testing with this on my local site, and it seemed to make a performance increase.

### CODE REVIEW

File: website/views/shared/tasks/index.jade [hunk 1]
```diff
@@ -6,7 +6,7 @@ include ./task_view/mixins
 script(id='templates/habitrpg-tasks.html', type="text/ng-template")
   .tasks-lists.container-fluid
     .row
-      .col-md-3.col-sm-6(ng-repeat='list in lists', ng-class='::{"rewards-module": list.type==="reward"}')
+      .col-md-3.col-sm-6(ng-repeat='list in lists track by $index', ng-class='::{"rewards-module": list.type==="reward"}')
```
Comment (crookedneighbor) [inline website/views/shared/tasks/index.jade:9]: Until the task migration is complete, we cannot use `track by $index` because some tasks have duplicate ids and Angular breaks if that happens while `track by` is turned on.

File: website/views/shared/tasks/task.jade [hunk 2]
```diff
@@ -1,5 +1,5 @@
 li(id='task-{{::task.id}}', 
-  ng-repeat='task in obj[list.type+"s"] | filterByTextAndNotes: obj.filterQuery | conditionalOrderBy: list.view=="dated":"date"', 
+  ng-repeat='task in obj[list.type+"s"] track by $index | filterByTextAndNotes: obj.filterQuery | conditionalOrderBy: list.view=="dated":"date"',
```
Comment (crookedneighbor) [inline website/views/shared/tasks/task.jade]: Whoops, I actually mean this one.

File: website/views/options/social/index.jade [hunk 3]
```diff
@@ -49,7 +49,7 @@ script(type='text/ng-template', id='partials/options.social.guilds.public.html')
   div.container-fluid(style='margin-bottom: 1.5em')
     input.form-control(type='text',ng-model='guildSearch', placeholder=env.t('search'))
   table.table.table-striped(style='clear:left;')
-    tr(ng-repeat='group in groups.public | filter:guildSearch track by group._id')
+    tr(ng-repeat='group in groups.public track by $index | filter:guildSearch track by group._id')
```
Comment (crookedneighbor) [inline website/views/options/social/index.jade]: Isn't this already being tracked by `group._id`?

File: website/views/shared/tasks/index.jade (hunk 1 above)
Comment (Andrux51) [inline website/views/shared/tasks/index.jade:9]: $index in this case should be entirely unrelated to the data model. It
exists gives the ngRepeat directive a way of distinguishing each node in
the dom and in a way that's unique to each list. "track by" has nothing to do with ordering.

I'll defer to your judgment on the tasks part, are there other places where it doesn't make
sense? I'll clean up the PR to target more specific places if so.

More info here: https://docs.angularjs.org/api/ng/directive/ngRepeat#tracking-and-duplicates

File: website/views/options/social/chat-message.jade [hunk 4]
```diff
@@ -1,6 +1,6 @@
 mixin chatMessages(inbox)
   ul.list-unstyled.tavern-chat
-    - var ngRepeat = inbox ? 'message in user.inbox.messages | toArray:true | orderBy:"sort":true' : 'message in group.chat track by message.id'
+    - var ngRepeat = inbox ? 'message in user.inbox.messages track by $index | toArray:true | orderBy:"sort":true' : 'message in group.chat track by message.id'
```
Comment (crookedneighbor) [inline website/views/options/social/chat-message.jade]: This causes unpredictable ordering for the messages in the inbox.

**Comment by Andrux51** [comment]
Thanks for catching the instances where there's already a track by filter. I was using sublime without the jade plugin so it was a little sketchy and I guess I got overzealous. I'll clean this up and re-commit.

File: website/views/options/social/chat-message.jade (hunk 4 above)
Comment (Andrux51) [inline website/views/options/social/chat-message.jade]: I think it's because I put the track by filter in the wrong part of the statement. Cleaning this up, then I would appreciate another look :)

File: website/views/options/social/index.jade (hunk 3 above)
Comment (Andrux51) [inline website/views/options/social/index.jade]: Yes, looks like I got a little overzealous with this one, cleaning it up.

File: website/views/options/social/chat-box.jade [hunk 5]
```diff
@@ -14,7 +14,7 @@ form.chat-form(ng-if='user.flags.communityGuidelinesAccepted' ng-submit='postCha
     textarea.form-control(rows=4, ui-keydown='{"meta-enter":"postChat(group,message.content)"}', ui-keypress='{13:"postChat(group,message.content)"}', ng-model='message.content', updateinterval='250', flag='@', at-user, auto-complete placeholder="{{group._id == 'habitrpg' ? env.t('tavernCommunityGuidelinesPlaceholder') : ''}}", ng-disabled='_sending == true')
     span.user-list
       ul.list-at-user(ng-show="query")
-        li(ng-repeat='msg in response | filter:filterUser | limitTo: 5', ng-click='performCompletion(msg)')
+        li(ng-repeat='msg in response track by $index | filter:filterUser | limitTo: 5', ng-click='performCompletion(msg)')
```
Comment (crookedneighbor) [inline website/views/options/social/chat-box.jade]: This is in the wrong spot and breaks the `limitTo`

screen shot 2015-06-26 at 10 49 50 pm

File: website/views/shared/modals/members.jade [hunk 6]
```diff
@@ -84,7 +84,7 @@ script(type='text/ng-template', id='modals/send-gift.html')
       .panel-heading=env.t('subscription')
       .panel-body
         .form-group
-          .radio(ng-repeat='block in Content.subscriptionBlocks | toArray | omit:"discount==true" | orderBy:"months"')
+          .radio(ng-repeat='block in Content.subscriptionBlocks track by $index | toArray | omit:"discount==true" | orderBy:"months"')
```
Comment (crookedneighbor) [inline website/views/shared/modals/members.jade]: I think the `track by` needs to be after the filters.

**Comment by Andrux51** [comment]
The latest commit should take care of the remaining places these were off. Sorry for having you help track them down, I hope this change helps the site performance!

File: website/views/main/filters.jade [hunk 7]
```diff
@@ -28,13 +28,13 @@
             button(ng-click='createTag()')=env.t('add')
       div(ng-if='!areTagsHidden || _editing')
         ul(ng-if='_editing')
-          li.filters-edit(ng-class='{active: user.filters[tag.id]}', ng-repeat='tag in user.tags', bindonce='user.tags')
+          li.filters-edit(ng-class='{active: user.filters[tag.id]}', ng-repeat='tag in user.tags track by $index', bindonce='user.tags')
```
Comment (crookedneighbor) [inline website/views/main/filters.jade]: I think this messes up the re-ordering of tags.

File: website/views/main/filters.jade [hunk 8]
```diff
@@ -28,13 +28,13 @@
...
             form.hrpg-input-group
               input(type='text', ng-model='tag.name', ui-keyup="{13: 'saveOrEdit()'}")
               button(type='button', ng-click='user.ops.deleteTag({params:{id:tag.id}})')
                 span.glyphicon.glyphicon-trash
         ul(ng-if='!_editing', hrpg-sort-tags)
-          li.filters-tags(ng-class='{active: user.filters[tag.id], challenge: tag.challenge}', ng-repeat='tag in user.tags', bindonce='user.tags')
+          li.filters-tags(ng-class='{active: user.filters[tag.id], challenge: tag.challenge}', ng-repeat='tag in user.tags track by $index', bindonce='user.tags')
```
Comment (crookedneighbor) [inline website/views/main/filters.jade]: Might just be this one. Not sure if it matters for the editting view.

**Comment by crookedneighbor** [comment]
Awesome, after the changes, I think everything is good, but I'm going to let another dev or two give some input.

**Comment by Alys** [comment]
@Andrux51 If you'd like to test this in an environment very similar to production, you can go to https://beta.habitrpg.com/ and log in with your normal account (or with a test account if you prefer). It uses the same database as the real HabitRPG site, so if you log in with your usual details, you will be using your real account.

Note that after the testing for this PR is finished, the beta site will be used for other purposes, and it is then not recommended that you use it at all because it could experimental code.

**Comment by Alys** [comment]
Test and other stuff for me to do:
- [ ] equipment page contains all gear
  - [x] all sub-sections are correct
  - [x] Battle Gear and Costume sides are identical
  - [x] gear handles equip/dequip correctly (both sides)
  - [ ] make note on that equipment sorting PR that this change needs to be merged and tested
- [x] seasonal shop equipment (delete all gear from test account)
  - [x] all gear is in correct order
  - [x] all gear can be bought for correct price
  - [x] gear still looks correct after some is bought
- [ ] seasonal shop quests are correct (fake this on a local install)
- [ ] time traveller equipment (delete all gear from test account)
  - [ ] all gear is in correct order
  - [ ] all gear can be bought for correct hourglass price
  - [ ] gear still looks correct after some is bought
- [ ] inventory eggs are correct, including after selling some (correct gold received?) and then all
- [ ] inventory hatching potions are correct, including after selling some (correct gold received?) and then all
- [ ] inventory foods are correct, including after selling some (correct gold received?) and then all
- [ ] inventory quests are correct, including after using one, and after deleting them all
- [ ] market eggs are correct and can be bought correctly
  - [ ] non-quest
  - [ ] quest; only from quests the account has completed (but note there's no `track by $index` for these
  - [ ] TRex quest eggs have correct layout and popup (both with and without quest completion) (but note there's no `track by $index` for these
- [ ] market hatching potions are correct and can be bought correctly
- [ ] market foods are correct and can be bought correctly
- [ ] market eggs are correct and can be bought correctly
- [ ] market quests are correct and can be bought correctly
  - [ ] quests cannot be bought when you haven't done the previous in the chain
- [ ] food appears correctly in tray on pets page, and layout adjusts correct after eating all of one food (try foods in both columns)
- [x] animal ears appear correctly on avatar page before any purchase, after some purchases, after all purchases
- [ ] webhooks on Settings API page are correct
- [ ] website/views/options/settings.jade partials/options.settings.subscription.html : subscription monthly blocks on Settings Subscription page are correct and on gem-in-toolbar page
- [ ] guild-specific challenges are correct in box on guild/Tavern page
- [ ] in partials/options.social.challenges.detail.close.html, closingChal.members are correct
- [ ] in partials/options.social.challenges.detail.html, challenge.members are correct
- [ ] in partials/options.social.challenges.html, 'group in groupsFilter' is correct
- [x] when typing a chat reply containing @, the list of matching users is as correct as it ever gets - guild, tavern, party, inbox?
- [x] inbox messages are correct, complete, and in the right order
- [ ] website/views/options/social/group.jade h4:invited : in a guild and a party, the list of invites is correct (zero, one, some; box is invisible when zero)
- [ ] partials/options.social.guilds.public.html : 'invitation in user.invitations.guilds' is correct
- [x] hall of heros is correct
- [x] hall of patrons is correct
- [ ] for active collection quest, the list of items collected is correct, for quests with one type of item and with two types 
- [ ] for inactive collection quest (website/views/options/social/quests/questNotActive.jade), the list of items to be collected is correct, for quests with one type of item and with two types 
- [x] in Tavern, list of staff and mods is correct with Alys in appropriate first place amongst mods and deilann in highly appropriate last place; profile clicks work correctly
- [x] party members in header are correct; test with a couple of sort option changes
- [ ] notification icon drop-down in toolbar shows all guilds you are invited to and clicking on them works
- [x] notification icon drop-down in toolbar shows all guilds you have new mesages for and clicking on them works - both to delete the notification and to visit the guild
- [ ] website/views/shared/modals/invite-friends.jade - 'email in emails' is correct
- [ ] website/views/shared/modals/members.jade modals/send-gift.html under subscription heading : subscription monthly blocks are correct
- [ ] partials/options.social.party.quest-rewards.html : quest rewards box has correct drop items (test with zero, one, more items)
- [ ] modals/showQuest.html : collection quest items to collect are correct (zero, one, two)
- [ ] modals/buyQuest.html : collection quest items to collect are correct (zero, one, two)
- [ ] modals/questInvitation.html : collection quest items to collect are correct (zero, one, two)
- [x] quest list in Achievements is correct
- [ ] website/views/shared/profiles/stats.jade h4:equipment : 
  - [ ] '(k,v) in profile.items.gear.equipped track by $index'
  - [ ] 'stat in ["str","con","per","int"] track by $index'
- [x] editing a task: checklist items are present and in the correct order and can be edited, deleted, added, moved
- [x] editing a task: tags are present and in the correct order and selected and deselected
- [ ] website/views/shared/tasks/index.jade task list : 'list in lists track by $index' and note [discussion](https://github.com/HabitRPG/habitrpg/pull/5500/files#r33408751
- [x] viewing a task: checklist items are present and in the correct order and can be ticked off and on
- [x] skills are correct - check a couple of classes
- [ ] rewards : `li.task.reward-item(ng-repeat='item in itemStore track by $index',popover-trigger='mouseenter', popover-placement='top', popover='{{item.key == "armoire" && !user.flags.armoireEmpty ? env.t("armoireNotesFull") + armoireCount(user.items.gear.owned) : item.notes()}}')`

**Comment by Andrux51** [comment]
@Alys I've done some testing on the beta site of lists that could potentially be affected by the change in this PR. It looks like there's a bit of performance increase on the guild chat, which was the original goal I had mind when making these changes.

chat ats

**Comment by Alys** [comment]
FYI I'm working through the tests now and am hoping to finish them today.

**Comment by Alys** [comment]
The following behaviours happen using the beta site (which contains this PR) but not using our staging site (production and production-ready code).

When editing a To-Do with a checklist, if you drag one of the checklist items downwards (not upwards) to change the order, then all of the items jump to different places. When you save the To-Do, the items go back to their correct locations, with the one you moved in its correct new place, however the wrong sort order when dragging makes sorting too difficult.

The first time I went to the Hall of Patrons, the entries were not sorted correctly. However a reload fixed that, and I haven't been able to replicate the problem. If no one else can replicate it or see conde that might have caused it, then we should probably consider that to be a once-off glitch.

On the Time Travelers page, if you buy the Winged Messenger Set by clicking on the first item in that set, that first item disappears but the other two do not (even though all three are added to your list of owned gear). If you try to buy other items, other odd behaviour occurs, such as the wrong items disappearing. After buying some of the items and reloading the page, you can get odd mixtures of items under the wrong headings:
screen shot 2015-07-11 at 4 44 25 pm

On the Inventory page, I reloaded it and confirmed that the count of eggs was correct, based on my account's data in the database. Then I modified the data in the database, reducing the value of one of the eggs to zero. Then I synced (not reloaded) the Inventory page. The egg counts were wrong - the egg that had been set to zero was still visible but had been given the value of the next egg in the list (and so on for the other eggs down the list). This exact circumstance won't happen in production use of course, but it mimics what will happen if the user has two tabs open and takes actions in one (e.g., hatching a pet), then syncs the other.

I'm going to stop testing now, despite my promise to finish this today. These problems will need to be fixed, and I think it's better that I do the other tests after that.

**Comment by Alys** [comment]
@Andrux51 The beta site no longer has this PR for testing, but much of the testing should be possible in a local install, and we'll be happy to set up the beta site again when needed.

**Comment by crookedneighbor** [comment]
@Andrux51 What's the state of this?

**Comment by Alys** [comment]
@Andrux51 I've been thinking that if you would like to proceed with this, it would be better to submit a PR containing just one or two of the changes at a time. That gives you and us time to do thorough testing for each change without having a huge test burden all at once. It also greatly reduces the chance of merge conflicts since the turnaround time would be much shorter.

Perhaps start with a couple of changes that you think will have the most impact.

**Comment by Andrux51** [comment]
@crookedneighbor @Alys 

Two things:

1.) I don't know if my account has enough data to QA some of the things mentioned in this discussion.

2.) In all honesty, I don't have time for now to do a site-wide QA pass for issues this may present. It really should have been a simple change that didn't affect the way data is displayed, but it seems like that hasn't been the case.

I think the suggestion to release it in smaller chunks is a good idea, and I may be able to make some commits with that in mind in the coming weeks/months, but nothing planned at the moment.

Basically what I'm trying to say is I'm washing my hands of this for the time being. If you want to close this PR, that's perfectly ok with me, or whatever makes the most sense.

**Comment by crookedneighbor** [comment]
@Andrux51 sounds good to me. I'm going to close this for now, but I'd encourage you to open a new PR for the a section that Alys tested that worked as expected. 

If I have a spare moment, I may do the same thing.

**Comment by Alys** [comment]
Note that I ticked off items on my checklist when they failed and I wrote a comment about them, as well as when they succeeded. A tick mark means "tested", not necessarily "working".
